projects/wizard_of_wikipedia/knowledge_retriever/knowledge_retriever.py

```diff
@@ -0,0 +1,341 @@
+        self.dialogue_history = []
+        self.checked_sentence_history = []
+
+    def _set_up_tfidf_retriever(self, opt):
+        retriever_opt = {
+            'model_file': opt['retriever_model_file'],
+            'remove_title': False,
+            'datapath': opt['datapath'],
+            'override': {'remove_title': False},
+        }
+        self.retriever = create_agent(retriever_opt)
+
+        self._set_up_sent_tok()
+        wiki_map_path = os.path.join(self.model_path, 'chosen_topic_to_passage.json')
+        self.wiki_map = json.load(open(wiki_map_path, 'r'))
+
+    def _set_up_selector(self, opt):
+        # make sure to download all relevant files
+        download(opt['datapath'])
+        selector_opt = {
+            'datapath': opt['datapath'],
+            'model_file': opt['selector_model_file'],
+            'eval_candidates': 'inline',
+            'model': 'transformer/biencoder',
+            'batchsize': 1,
+            'override': {'model': 'transformer/biencoder', 'batchsize': 1},
+        }
+        for k, v in self.opt.items():
+            if k not in selector_opt:
+                selector_opt[k] = v
+        self.selector = create_agent(selector_opt)
+
+    def _set_up_sent_tok(self):
+        try:
+            import nltk
+        except ImportError:
+            raise ImportError('Please install nltk (e.g. pip install nltk).')
+        # nltk-specific setup
+        st_path = 'tokenizers/punkt/{0}.pickle'.format('english')
+        try:
+            self.sent_tok = nltk.data.load(st_path)
+        except LookupError:
+            nltk.download('punkt')
+            self.sent_tok = nltk.data.load(st_path)
+
+    def maintain_retriever_history(self, obs, actor_id='apprentice'):
+        """
+        Maintain texts retrieved by the retriever to mimic the set-up
+        from the data collection for the task.
+        """
+
+        if actor_id == 'apprentice':
+            if self.retriever_history['episode_done']:
+                # clear history
+                self.retriever_history['chosen_topic'] = ''
+                self.retriever_history['wizard'] = ''
+                self.retriever_history['apprentice'] = ''
+                self.retriever_history['episode_done'] = False
+                self.dialogue_history = []
+
+            # save chosen topic
+            if 'chosen_topic' in obs:
+                self.retriever_history['chosen_topic'] = obs['chosen_topic']
+            if 'text' in obs:
+                self.retriever_history['apprentice'] = obs['text']
+
+            self.retriever_history['episode_done'] = obs['episode_done']
+
+        elif not self.retriever_history['episode_done']:
+            # only add the wizard item to the history if episode_done is not
+            # True
+            self.retriever_history['wizard'] = obs['text']
+
+        self.dialogue_history.append({'id': actor_id, 'text': obs['text']})
+
+    def get_chosen_topic_passages(self, chosen_topic):
```

Comment:
is this logic from the mturk task? looks familiar lol

Reply:
it's similar. it's been removed from the evaluation task. but it presents the text to the model agent slightly differently from how it was presented to the turk agent